### ISSUE TITLE
pseudoArtist - `withdrawAllUnlockedToken` will run OOG if there are too many locks due to unecessary while loop

### ISSUE THREAD

**Issue by sherlock-admin4** [sherlock-audit/2024-05-gamma-staking-judging#277]
pseudoArtist

medium

# `withdrawAllUnlockedToken` will run OOG if there are too many locks due to unecessary while loop

## Summary
The function `withdrawAllUnlockedToken()` in `Lock.sol` will run OOG in case there are too many locks due to unecessary concept of limit and while loop.
## Vulnerability Detail
`withdrawAllUnlockedToken` is used to  Withdraws all currently unlocked tokens where the unlock time has passed for the calling user, 
now initially it is fetching the total no of locks for the caller from ` uint256 lockCount = locklist.lockCount(msg.sender);` and initialising the `limit` and `page` to 0, and then running the while loop till the `limit < lockCount` , Inside the while loop it is getting the page of the locks for the user, and then running a for loop till `i < lockedBals.length`,  

Now the  problem is inside the for loop it is  removing all the locks from the list for which the lock time has passed, and all the locks will be removed inside the for loop and then the limit will be increased by 10 ` limit += 10;`, since all the locks are already removed there is no need to run the while loop again, but it will once again enter the while loop and keep on doing it until the limit exceeds the `lockCount`.

If there are too many locks for the user which is unclocked, this unecessary running of loop will cause excessive usage of gas and the function will run out of gas before transferring the total unlocked amount to the user which is done only after the while loop is executed.

` IERC20(stakingToken).safeTransfer(msg.sender, totalUnlocked);`

This will impact the function `withdrawAllUnlockedToken` and the function will run OOG.
## Code Snippet
https://github.com/sherlock-audit/2024-05-gamma-staking/blob/main/StakingV2/src/Lock.sol#L409-L430
```solidity
  function withdrawAllUnlockedToken() external override nonReentrant {
        uint256 lockCount = locklist.lockCount(msg.sender); 
        uint256 page;
        uint256 limit;
        uint256 totalUnlocked;

        while (limit < lockCount) {
            LockedBalance[] memory lockedBals = locklist.getLocks(msg.sender, page, lockCount); 
            for (uint256 i = 0; i < lockedBals.length; i++) {
                if (lockedBals[i].unlockTime != 0 && lockedBals[i].unlockTime < block.timestamp) {
                    totalUnlocked += lockedBals[i].amount;
                    locklist.removeFromList(msg.sender, lockedBals[i].lockId); 
                }
            }

            limit += 10; // Moves to the next page of locks.
            page++;
        }

        IERC20(stakingToken).safeTransfer(msg.sender, totalUnlocked); 
        emit WithdrawAllUnlocked(msg.sender, totalUnlocked); 
    }
```
## Tool used

Manual Review

## Recommendation
Remove the while loop and concept of limit from the above mentioned function

```solidity
 function withdrawAllUnlockedToken() external override nonReentrant {
        uint256 lockCount = locklist.lockCount(msg.sender); // Fetch the total number of locks for the caller.
        uint256 totalUnlocked;
            LockedBalance[] memory lockedBals = locklist.getLocks(msg.sender, page, lockCount); // Retrieves a page of locks for the user.
            for (uint256 i = 0; i < lockedBals.length; i++) {
                if (lockedBals[i].unlockTime != 0 && lockedBals[i].unlockTime < block.timestamp) {
                    totalUnlocked += lockedBals[i].amount; // Adds up the amount from all unlocked balances.
                    locklist.removeFromList(msg.sender, lockedBals[i].lockId); // Removes the lock from the list.
                }
            }
        IERC20(stakingToken).safeTransfer(msg.sender, totalUnlocked); // Transfers the total unlocked amount to the user.
        emit WithdrawAllUnlocked(msg.sender, totalUnlocked); // Emits an event logging the withdrawal.
    }
 ```

Duplicate of #214
